ebpf_ffi/cbpf.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ebpf_ffi/cbpf.h"
+
+#include <linux/bpf.h>
+#include <linux/filter.h>
+#include <linux/if_ether.h>
+#include <linux/kernel.h>
+#include <netpacket/packet.h>
+
+namespace cbpf_ffi {
+
+// This constant was determined arbitrarily, the number of 0's has incremented
+// when the size was no longer enough for the verifier logs.
+constexpr size_t kLogBuffSize = 100000000;
+}  // namespace cbpf_ffi
+
+void load_cbpf_program(void *prog_buff, std::string *error, int *socks) {
+  if (socketpair(AF_UNIX, SOCK_DGRAM, 0, socks) != 0) {
+    *error = strerror(errno);
+  }
+
+  // cBPF programs have two relevant structures: sock_filter, and sock_fprog
+  // https://www.kernel.org/doc/html/latest/networking/filter.html#structure
+  struct sock_filter *insn = (struct sock_filter *)prog_buff;
+  struct sock_fprog program;
+  int size = sizeof(insn);
```

Comment:
could we inline `sizeof(insn)` in the next line instead of putting it in a variable here?

Reply:
Unfortunately doing so gives the following warning:
```
ebpf_ffi/cbpf.cc:33:30: warning: division 'sizeof (sock_filter*) / sizeof (sock_filter)' does not compute the number of array elements [-Wsizeof-pointer-div]
   33 |   program.len = sizeof(insn) / sizeof(insn[0]);
      |                 ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
ebpf_ffi/cbpf.cc:30:23: note: first 'sizeof' operand was declared here
   30 |   struct sock_filter *insn = (struct sock_filter *)prog_buff;
 ```